README.md
Utility classes and functions for comfortable and "safe" functional programming.

## Usage

A simple usage example:

```dart
import 'package:funcy/funcy.dart';
import 'package:meta/meta.dart';

void main() {
  final employeeJson = retrieveEmployeeJson();
  final maybeEmployee =
      Employee.tryFromJson(employeeJson) | tryFallbackEmployee();
  maybeEmployee.option(
    () => print('Something really bad is happening '
        'even fallback employee does not work.'),
    (employee) => print('${employee.name} is ${employee.age}.'),
  );
}

Map<String, dynamic> retrieveEmployeeJson() => <String, dynamic>{
      'name': 'Bob',
      'position': 'boss',
      'age': 42,
    };

Option<Employee> tryFallbackEmployee() => Some(
      Employee(
        name: 'Fallback Employee',
        position: Position.subsubboss,
        age: -1,
      ),
    );

enum Position {
  boss,
  subboss,
  subsubboss,
}

Option<Position> tryParsePosition(String positionStr) {
  switch (positionStr) {
    case 'boss':
      return const Some(Position.boss);
    case 'subboss':
      return const Some(Position.subboss);
    case 'subsubboss':
      return const Some(Position.subsubboss);
    default:
      return const None();
  }
}

class Employee {
  Employee({
    @required this.name,
    @required this.position,
    @required this.age,
  });

  final String name;
  final Position position;
  final int age;

  static Option<Employee> tryFromJson(Map<String, dynamic> json) {
    return json.tryGet<String>('name').bind(
      (name) {
        return json.tryGet<String>('position').bind(tryParsePosition).bind(
          (position) {
            return json.tryGet<int>('age').guard(_ageTest).map(
              (age) {
                return Employee(
                  name: name,
                  position: position,
                  age: age,
                );
              },
            );
          },
        );
      },
    );
  }

  static bool _ageTest(int age) => age >= 18;
}
```

CHANGELOG.md
## 3.1.0

- Fix broken generic parameters in datatype conversion extensions.

- Add [NullableToOption] extension with [asOptional()] to convert nullable
(actually, any) types to [Option].

## 3.0.0

- Make [Future] transformer extensions actually useful.

- Add methods and extensions for easy conversion between datatypes.

## 2.1.0

- Add [Loadable.isSuccess] and [Loadable.dataOr].

## 2.0.1

- Fix bug with [JsonOptionExtension.tryGet] when called with [dynamic] 
as a type parameter it returned [Some] even when there where no proper value.

## 2.0.0

- Added [Failed.failure] field - you very often need a reason of failure.
So, [Loadable] is now parameterised by 2 generic types
and is more resembling of [Either] with extra [Loading] case.
That is also the reason [Loadable] does not implement [Alternative] anymore.

## 1.0.1

- Lowered *meta* package minimum version to support current version of Flutter.

## 1.0.0

- Initial version

test/datatypes/either/either_test.dart
import 'package:funcy/funcy.dart';
import 'package:test/test.dart';

void main() {
  const rightValue = 7;
  final right = Either.right<String, int>(rightValue);
  const leftValue = 'seven';
  final left = Either.left<String, int>(leftValue);

  group(
    '*fromOption*',
    () {
      test(
        'if [optionalRight] is Some(right) - returns Right(right)',
        () {
          const Option<int> someRight = Some(rightValue);
          expect(Either.fromOption(someRight, leftValue), right);
        },
      );

      test(
        'if [optionalRight] is None() - returns Left([left])',
        () {
          const Option<int> noneRight = None();
          expect(Either.fromOption(noneRight, leftValue), left);
        },
      );

      test(
        'if [optionalRight] is null - throws ArgumentError',
        () {
          const Option<int> nullOptionalRight = null;
          expect(
            () => Either.fromOption(nullOptionalRight, leftValue),
            throwsArgumentError,
          );
        },
      );
    },
  );

  group(
    '*fromLoaded*',
    () {
      test(
        'if [loaded] is Failed(left) - returns Left(left)',
        () {
          const Loaded<String, int> failed = Failed(leftValue);
          expect(Either.fromLoaded(failed), left);
        },
      );

      test(
        'if [loaded] is Success(right) - returns Right(right)',
        () {
          const Loaded<String, int> success = Success(rightValue);
          expect(Either.fromLoaded(success), right);
        },
      );

      test(
        'if [loaded] is null - throws ArgumentError',
        () {
          const Loaded<String, int> nullLoaded = null;
          expect(() => Either.fromLoaded(nullLoaded), throwsArgumentError);
        },
      );
    },
  );

  group(
    '*fromLoadable*',
    () {
      group(
        'if [loadable] is',
        () {
          test(
            'Failed(left) - returns Left(left)',
            () {
              const Loaded<String, int> failed = Failed(leftValue);
              expect(Either.fromLoaded(failed), left);
            },
          );

          test(
            'Success(right) - returns Right(right)',
            () {
              const Loaded<String, int> success = Success(rightValue);
              expect(Either.fromLoaded(success), right);
            },
          );

          test(
            'null - throws ArgumentError',
            () {
              const Loaded<String, int> nullLoaded = null;
              expect(() => Either.fromLoaded(nullLoaded), throwsArgumentError);
            },
          );
        },
      );
    },
  );

  group(
    '*mirrored*',
    () {
      test(
        'if Left(left) - returns Right(left)',
        () {
          expect(left.mirrored(), const Right<int, String>(leftValue));
        },
      );
      test('if Right(right) - returns Left(right)', () {
        expect(right.mirrored(), const Left<int, String>(rightValue));
      });
    },
  );

  group('equality', () {
    group('is true for', () {
      test('Rights with equal values', () {
        const value = 7;
        const right1 = const Right<String, int>(value);
        const right2 = const Right<String, int>(value);
        expect(right1 == right2, isTrue);
      });

      test('Right and Either that is Right with equal values', () {
        const value = 7;
        const right = const Right<String, int>(value);
        final either = Either.right<String, int>(value);
        expect(right == either, isTrue);
      });

      test('Eithers that are Rights with equal values', () {
        const value = 7;
        final either1 = Either.right<String, int>(value);
        final either2 = Either.right<String, int>(value);
        expect(either1 == either2, isTrue);
      });

      test('Lefts with equal values', () {
        const value = 7;
        const left1 = const Left<int, String>(value);
        const left2 = const Left<int, String>(value);
        expect(left1 == left2, isTrue);
      });

      test('Left and Either that is Left with equal values', () {
        const value = 7;
        const left = const Left<int, String>(value);
        final either = Either.left<int, String>(value);
        expect(left == either, isTrue);
      });

      test('Eithers that are Lefts with equal values', () {
        const value = 7;
        final either1 = Either.left<int, String>(value);
        final either2 = Either.left<int, String>(value);
        expect(either1 == either2, isTrue);
      });
    });

    group('is false for', () {
      test('Rights with non-equal values', () {
        const right1 = const Right<String, int>(7);
        const right2 = const Right<String, int>(0);
        expect(right1 == right2, isFalse);
      });

      test('Lefts with non-equal values', () {
        const left1 = const Left<int, String>(7);
        const left2 = const Left<int, String>(0);
        expect(left1 == left2, isFalse);
      });

      test('Eithers that are Left and Right with any values', () {
        final left = Either.left<int, int>(7);
        final right = Either.right<int, int>(7);
        expect(left == right, isFalse);
      });
    });
  });

  group('*isLeft* is', () {
    test('true, when Left.', () {
      const Either<String, int> strOrInt = const Left('seven');
      expect(strOrInt.isLeft, isTrue);
    });

    test('false, when Right.', () {
      const Either<String, int> strOrInt = const Right(7);
      expect(strOrInt.isLeft, isFalse);
    });
  });

  group('*isRight* is', () {
    test('false, when Left.', () {
      const Either<String, int> strOrInt = const Left('seven');
      expect(strOrInt.isRight, isFalse);
    });

    test('true, when Right.', () {
      const Either<String, int> strOrInt = const Right(7);
      expect(strOrInt.isRight, isTrue);
    });
  });

  group('*map*', () {
    group('if Right', () {
      test('applies [f] to value and returns it in Right', () {
        final f = (int n) => n - 2;
        final rightFive = right.map(f);
        expect(rightFive, Right<String, int>(f(rightValue)));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => right.map<bool>(null), throwsArgumentError);
      });
    });

    group('if Left', () {
      test('does not change Left value', () {
        final f = (int n) => n / 2;
        final leftUpd = left.map(f);
        expect(leftUpd, const Left<String, double>(leftValue));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => left.map<bool>(null), throwsArgumentError);
      });
    });
  });

  group('*bind*', () {
    group('if Right', () {
      test('applies [f] to value', () {
        final f = (int n) => Either.right<String, bool>(n > 0);
        final result = right.bind(f);
        expect(result, f(rightValue));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => right.bind<bool>(null), throwsArgumentError);
      });
    });

    group('if Left', () {
      test('does not change Left value', () {
        final f = (int n) => Either.right<String, double>(n / 2);
        final leftUpd = left.bind(f);
        expect(leftUpd, const Left<String, double>(leftValue));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => left.bind<bool>(null), throwsArgumentError);
      });
    });
  });

  group('*either*', () {
    group('if Right', () {
      test('applies [fromRight] to value', () {
        final result = right.either(identity, (n) => n.toString());
        expect(result, rightValue.toString());
      });

      test('and [fromLeft] is null throws ArgumentError', () {
        expect(() => right.either(null, identity), throwsArgumentError);
      });

      test('and [fromRight] is null throws ArgumentError', () {
        expect(() => right.either(identity, null), throwsArgumentError);
      });
    });

    group('if Left', () {
      test('applies [fromLeft] to value', () {
        final result = left.either((s) => s.length, identity);
        expect(result, leftValue.length);
      });

      test('and [fromLeft] is null throws ArgumentError', () {
        expect(() => left.either(null, identity), throwsArgumentError);
      });

      test('and [fromRight] is null throws ArgumentError', () {
        expect(() => left.either(identity, null), throwsArgumentError);
      });
    });
  });

  group('*fromRight*', () {
    group('if Right', () {
      test('applies [fromRight] to value', () {
        final f = (int n) => n * 2;
        final result = right.fromRight(null, f);
        expect(result, f(rightValue));
      });

      test('and [fromRight] is null throws ArgumentError', () {
        expect(() => right.fromRight(7, null), throwsArgumentError);
      });
    });

    group('if Left', () {
      const value = 77;

      test('returns [ifLeft]', () {
        final result = left.fromRight(value, identity);
        expect(result, value);
      });

      test('and [fromRight] is null throws ArgumentError', () {
        expect(() => left.fromRight(value, null), throwsArgumentError);
      });
    });
  });

  group('*rightOr*', () {
    group('if Right', () {
      test('returns value', () {
        expect(right.rightOr(null), rightValue);
      });
    });
    group('if Left', () {
      test('returns [ifLeft]', () {
        const value = 77;
        expect(left.rightOr(value), value);
      });
    });
  });

  group('*fromLeft*', () {
    group('if Right', () {
      const value = 'seventy seven';

      test('returns [ifRight]', () {
        final result = right.fromLeft(value, identity);
        expect(result, value);
      });

      test('and [fromLeft] is null throws ArgumentError', () {
        expect(() => right.fromLeft(value, null), throwsArgumentError);
      });
    });

    group('if Left', () {
      test('applies [fromLeft] to value', () {
        final f = (String s) => s.length;
        final result = left.fromLeft(null, f);
        expect(result, f(leftValue));
      });

      test('and [fromLeft] is null throws ArgumentError', () {
        expect(() => left.fromLeft(7, null), throwsArgumentError);
      });
    });
  });

  group('*leftOr*', () {
    group('if Right', () {
      test('returns [ifRight]', () {
        const value = 'seventy seven';
        expect(right.leftOr(value), value);
      });
    });

    group('if Left', () {
      test('returns value', () {
        expect(left.leftOr(null), leftValue);
      });
    });
  });

  group('*toOption*', () {
    group('if Right', () {
      test('returns Some with value', () {
        expect(right.toOption(), const Some(rightValue));
      });
    });

    group('if Left', () {
      test('returns None', () {
        expect(left.toOption(), const None<int>());
      });
    });
  });
}


test/datatypes/option/option_test.dart
import 'package:funcy/funcy.dart';
import 'package:test/test.dart';

void main() {
  const value = 7;
  final some = Option.fromNullable(value);
  final none = Option.fromNullable<int>(null);

  group('*equality*', () {
    group('is true for', () {
      test('Somes with equal values', () {
        const some1 = const Some(7);
        const some2 = const Some(7);
        expect(some1 == some2, isTrue);
      });

      test('Some and Option that is Some with equal values', () {
        const some = const Some(7);
        final option = Option.fromNullable(7);
        expect(some == option, isTrue);
      });

      test('Nones', () {
        const none1 = const None<int>();
        const none2 = const None<int>();
        expect(none1 == none2, isTrue);
      });

      test('None and Option that is None', () {
        const none = const None<int>();
        final option = Option.fromNullable<int>(null);
        expect(none == option, isTrue);
      });
    });

    group('is false for', () {
      test('Options that are Some and None', () {
        final some = Option.fromNullable(7);
        final none = Option.fromNullable<int>(null);
        expect(some == none, isFalse);
      });
    });
  });

  group('*Option.fromNullable*', () {
    group('if [nullableValue] is', () {
      test('not null returns Some with [nullableValue]', () {
        const value = 7;
        final some = Option.fromNullable(value);
        expect(some, const Some(value));
      });

      test('null returns None', () {
        expect(Option.fromNullable<int>(null), const None<int>());
      });
    });
  });

  group('*Option.fromEither*', () {
    group('if [either] is', () {
      test('Right returns Some with right value', () {
        const value = 'seven';
        final right = Either.right<int, String>(value);
        final some = Option.fromEither(right);
        expect(some, const Some(value));
      });

      test('Left returns None', () {
        const value = 7;
        final left = Either.left<int, String>(value);
        expect(Option.fromEither(left), const None<String>());
      });
    });
  });

  group('*Option.fromLeftEither*', () {
    group('if [either] is', () {
      test('Left returns Some with left value', () {
        const value = 7;
        final left = Either.left<int, String>(value);
        final some = Option.fromLeftEither(left);
        expect(some, const Some(value));
      });

      test('Right returns None', () {
        const value = 'seven';
        final right = Either.right<int, String>(value);
        expect(Option.fromLeftEither(right), const None<int>());
      });
    });
  });

  group('*isSome*', () {
    test('if is Some returns true', () {
      expect(some.isSome, isTrue);
    });

    test('if is None returns false', () {
      expect(none.isSome, isFalse);
    });
  });

  group('*isNone', () {
    test('if is Some returns false', () {
      expect(some.isNone, isFalse);
    });

    test('if is None returns true', () {
      expect(none.isNone, isTrue);
    });
  });

  group('*valueOr*', () {
    test('if is Some returns value', () {
      expect(some.valueOr(null), value);
    });

    test('if is None returns [ifNone]', () {
      const ifNone = 77;
      expect(none.valueOr(ifNone), ifNone);
    });
  });

  group('*valueOrDo*', () {
    group('if is Some', () {
      test('returns value', () {
        expect(some.valueOrDo(() => value - 1), value);
      });

      test('and [ifNone] is null throws ArgumentError', () {
        expect(() => some.valueOrDo(null), throwsArgumentError);
      });
    });

    group('if is None', () {
      test('calls [ifNone]', () {
        final ifNone = () => 77;
        expect(none.valueOrDo(ifNone), ifNone());
      });

      test('and [ifNone] is null throws ArgumentError', () {
        expect(() => none.valueOrDo(null), throwsArgumentError);
      });
    });
  });

  group('*option*', () {
    group('if is Some', () {
      test('applies [f] to value', () {
        final f = (int n) => n * 2;
        expect(some.option(constant(null), f), f(value));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => some.option(constant(77), null), throwsArgumentError);
      });
    });

    group('if is None', () {
      test('calls [ifNone]', () {
        final ifNone = () => 'seven';
        expect(none.option(ifNone, (n) => n.toString()), ifNone());
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => none.option(constant(77), null), throwsArgumentError);
      });
    });
  });

  group('*doIfSome*', () {
    var counter = 0;
    final action = (int n) => counter += n;

    setUp(() {
      counter = 0;
    });

    group('if is Some', () {
      test('calls [action] with value', () {
        final initial = counter;
        some.doIfSome(action);
        final result = counter;
        counter = initial;
        action(value);
        expect(counter, result);
      });

      test('and [action] is null throws ArgumentError', () {
        expect(() => some.doIfSome(null), throwsArgumentError);
      });
    });

    group('if is None', () {
      test('does nothing', () {
        final initial = counter;
        none.doIfSome(action);
        final result = counter;
        expect(initial, result);
      });

      test('and [action] is null throws ArgumentError', () {
        expect(() => none.doIfSome(null), throwsArgumentError);
      });
    });
  });

  group('*map*', () {
    final f = (int n) => n.isOdd;

    group('if is Some', () {
      test('applies [f] to value and returns the result in Some', () {
        final result = some.map(f);
        expect(result, Some(f(value)));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => some.map<int>(null), throwsArgumentError);
      });
    });

    group('if is None', () {
      test('returns None', () {
        final result = none.map(f);
        expect(result, const None<bool>());
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => none.map<int>(null), throwsArgumentError);
      });
    });
  });

  group('*bind*', () {
    final f = (int n) => n.isOdd ? Some('odd: $n') : const None<String>();
    group('if is Some', () {
      test('applies [f] to value', () {
        final result = some.bind(f);
        expect(result, f(value));
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => some.bind<int>(null), throwsArgumentError);
      });
    });

    group('if is None', () {
      test('returns None', () {
        final result = none.bind(f);
        expect(result, const None<String>());
      });

      test('and [f] is null throws ArgumentError', () {
        expect(() => none.bind<int>(null), throwsArgumentError);
      });
    });
  });

  group('*or*', () {
    group('if is Some', () {
      test('returns self', () {
        const Option<int> another = null;
        final some2 = some.or(another);
        expect(some, some2);
      });
    });

    group('if is None', () {
      test('returns [another]', () {
        const another = const Some(7);
        final another2 = none.or(another);
        expect(another, another2);
      });
    });
  });

  group('*|*', () {
    group('if is Some', () {
      test('returns self', () {
        const Option<int> another = null;
        final some2 = some | another;
        expect(some, some2);
      });
    });

    group('if is None', () {
      test('returns [another]', () {
        const another = const Some(7);
        final another2 = none | another;
        expect(another, another2);
      });
    });
  });

  group('guard', () {
    group('if is Some and predicate is', () {
      test('true returns Some', () {
        expect(some.guard((_) => true), some);
      });

      test('false returns None', () {
        expect(some.guard((_) => false), none);
      });

      test('null throws ArgumentError', () {
        expect(() => some.guard(null), throwsArgumentError);
      });
    });

    group('if is None and predicate is', () {
      test('true returns None', () {
        expect(none.guard((_) => true), none);
      });

      test('false returns None', () {
        expect(none.guard((_) => false), none);
      });

      test('null throws ArgumentError', () {
        expect(() => none.guard(null), throwsArgumentError);
      });
    });
  });

  group('JoinOptionExtension:', () {
    group('*join*', () {
      test('if is Some returns inner Option', () {
        final innerOption = Option.fromNullable(7);
        final someOption = Option.fromNullable(innerOption);
        expect(someOption.join(), innerOption);
      });

      test('if is None returns None', () {
        final none = Option.fromNullable<Option<int>>(null);
        expect(none.join(), const None<int>());
      });
    });
  });
}


test/utils/function/function_test.dart
import 'package:test/test.dart';

import 'package:funcy/funcy.dart';

void main() {
  group('*identity*', () {
    group('returns', () {
      test('the same value it is given', () {
        const value = 7;
        expect(identity(value), value);
      });
      test('null if given null', () {
        expect(identity(null), null);
      });
    });
  });

  group('*constFun*', () {
    test('creates function that always returns [value] whatever argument is',
        () {
      const value = 7;
      expect(constFun<int, int>(value)(10), value);
      expect(constFun<int, int>(value)(null), value);
    });
    test('returns null if [value] is null', () {
      expect(constFun<int, int>(null)(10), null);
    });
  });

  group('*constant*', () {
    test('creates function that always returns [value]', () {
      const value = 7;
      expect(constant(value)(), value);
    });

    test('returns null if [value] is null', () {
      expect(constant(null)(), null);
    });
  });

  group('FunctionFunctionExtension:', () {
    group('*then*', () {
      test(
          'creates function that applies [next] to the result '
          'of applying this to given argument', () {
        final x2 = (int n) => n * 2;
        final add3 = (int n) => n + 3;
        final str = (int n) => 'int: $n';
        final check = x2.then(add3).then(str);
        final match = (int n) => str(add3(x2(n)));
        expect(check(5), match(5));
      });

      test('throws ArgumentError if [next] is null', () {
        final x2 = (int n) => n * 2;
        expect(() => x2.then<bool>(null), throwsArgumentError);
      });
    });

    group('*thenConst*', () {
      test(
          'creates function that returns [value] '
          'after applying this to given argument', () {
        var counter = 1;
        final doubleCounter = (int n) => counter *= 2;
        final complete = doubleCounter
            .then(doubleCounter)
            .then(doubleCounter)
            .thenConst('complete');
        expect(complete(1), 'complete');
        expect(counter, 8);
      });
    });

    group('*after*', () {
      test(
          'creates function that applies this to the result '
          'of applying [first] to given argument', () {
        final x2 = (num n) => n * 2;
        final add3 = (num n) => n + 3;
        final str = (num n) => 'int: $n';
        final check = str.after(add3).after(x2);
        final match = (num n) => str(add3(x2(n)));
        expect(check(5), match(5));
      });

      test('throws ArgumentError if [first] is null', () {
        final x2 = (num n) => n * 2;
        expect(() => x2.after<bool>(null), throwsArgumentError);
      });
    });
  });

  group('FunctionObjectExtension:', () {
    group('*pipe*', () {
      test('creates function that applies [f] to this', () {
        final x2 = (int n) => n * 2;
        const value = 7;
        expect(value.pipe(x2), x2(value));
      });

      test('throws ArgumentError if [f] is null', () {
        const value = 7;
        expect(() => value.pipe<bool>(null), throwsArgumentError);
      });
    });
    group('*discardWith*', () {
      test(
          'creates function that applies this to '
          'given argument and returns [value]', () {
        var counter = 1;
        final doubleCounter = (int n) => counter *= 2;
        final complete = doubleCounter
            .then(doubleCounter)
            .then(doubleCounter)(1)
            .discardWith('complete');
        expect(complete, 'complete');
        expect(counter, 8);
      });
    });
  });
}


lib/src/utils/function.dart
/// Identity function - returns its argument.
/// ```
/// identity(7) // 7
/// identity(0) // 0
/// ```
/// Synonym for `(x) => x`.
/// Useful as a 'tear-off' (curried function),
/// where function is required as an argument.
/// ```
/// final optName = json.tryGet<String>('name');
/// optName.option(getDefaultName, identity);
/// ```
T identity<T>(T value) => value;

/// Creates constant function that always returns [value]
/// whatever given argument is.
/// ```
/// final const7 = constFun(7);
/// assertconst7(5) == 7);
/// assert(const7(0) == 7);
/// ```
/// Synonym for `(_) => value`. Useful as a 'tear-off' (curried function),
/// where function is required as an argument.
/// ```
/// final defaultValue = getDefaultValue();
/// List.generate(100, constFun(defaultValue));
/// ```
C Function(T) constFun<C, T>(C value) => (_) => value;

/// Creates constant function that always returns [value].
/// ```
/// final const7 = constant(7);
/// assert(const7() == 7);
/// assert(const7() == 7);
/// ```
/// Synonym for `() => value`. Useful as a 'tear-off' (curried function),
/// where function is required as an argument.
/// ```
/// final defaultLength = await getDefaultLength();
/// optString.option(constant(defaultLength), (s) => s.length);
/// optString.option(() => defaultLength, (s) => s.length);
/// ```
C Function() constant<C>(C value) => () => value;

/// Auxilary methods for function composition and *'fluent'* APIs.
extension FunctionFunctionExtension<A, B> on B Function(A) {
  /// Applies the function to given argument, then applies [next] to the result.
  ///
  /// ```
  /// final x2 = (n) => n * 2;
  /// final add3 = (n) => n + 3;
  /// x2.then(add3).then(print)(5); // prints '13'
  /// // is the same as
  /// print(add3(x2(5))); // prints '13'
  /// ```
  ///
  /// Throws [ArgumentError] if [next] is null.
  C Function(A) then<C>(C Function(B) next) {
    ArgumentError.checkNotNull(next, 'next');
    return (argument) => next(this(argument));
  }

  /// Applies the function to givent argument, then returns [value].
  ///
  /// Shorthand for `func.then(constant(value))`.
  C Function(A) thenConst<C>(C value) => then(constFun(value));

  /// Applies the function to the result of applying [first] to given argument.
  ///
  /// ```
  /// final x2 = (n) => n * 2;
  /// final add3 = (n) => n + 3;
  /// print.after(add3).after(x2)(5); // prints '13'
  /// // is the same as
  /// print(add3(x2(5))); // prints '13'
  /// ```
  B Function(C) after<C>(A Function(C) first) {
    ArgumentError.checkNotNull(first, 'first');
    return (argumentToFirst) => this(first(argumentToFirst));
  }
}

/// Auxilary methods for objects in function compositions and *'fluent'* APIs.
extension FunctionObjectExtension<T> on T {
  /// Applies [f] to the object.
  ///
  /// Can be useful in function compositions  and *'fluent'* APIs.
  /// ```
  /// List.generate(100, (i) => i * i)
  ///   .map((n) => createBussinessObject(n))
  ///   .where(bussinessFilter)
  ///   .take(7)
  ///   .toList(growable: false)
  ///   .pipe((objects) => usingBussinessObjects(objects));
  ///
  /// // is the same as
  ///
  /// usingBussinessObjects(
  ///   List.generate(100, (i) => i * i)
  ///     .map((n) => createBussinessObject(n))
  ///     .where(bussinessFilter)
  ///     .take(7)
  ///     .toList(growable: false));
  /// ```
  R pipe<R>(R Function(T) f) {
    ArgumentError.checkNotNull(f, 'f');
    return f(this);
  }

  /// Drops the object and returns [value].
  ///
  /// Shorthand for `obj.pipe((_) => anotherObj)`.
  ///
  /// ```
  /// final trace = (msg, value) => print('$msg: $value').discardWith(value);
  /// final seven = trace('Seven', 3 + 4); // prints 'Seven: 7'
  /// assert(seven, 7);
  /// ```
  R discardWith<R>(R value) => value;
}


lib/src/typeclasses/functor.dart
/// Functor typeclass.
///
/// Google *'typeclassopedia'*, if you are interested ^)
///
/// Basically, a box with magic sauce.
abstract class Functor<T> {
  /// Functor map.
  Functor<R> map<R>(R Function(T value) f);
}

/// Monad typeclass.
///
/// Google *'typeclassopedia'*, if you are interested ^)
///
/// Cooler functor. A box with magic sauce
/// that can be combined with other such boxes.
abstract class Monad<T> extends Functor<T> {
  /// Monad bind.
  Monad<R> bind<R>(covariant Monad<R> Function(T value) f);
}

/// Alternative typeclass.
///
/// Google *'typeclassopedia'*, if you are interested ^)
///
/// A functor that has a concept of choice between "success" and "failure".
abstract class Alternative<T> extends Functor<T> {
  /// Alternative choice.
  ///
  /// If it is "success" - returns itself. Otherwise, returns [another].
  Alternative<T> or(covariant Alternative<T> another);

  /// Alternative choice. [or] as an operator.
  ///
  /// If left argument is "success" - returns left argument.
  /// Otherwise, returns right argument.
  Alternative<T> operator |(covariant Alternative<T> another) => or(another);

  /// Alternative guard. If predicate returns false,
  /// returns a value representing *failure*.
  Alternative<T> guard(bool Function(T value) predicate);
}


lib/src/utils/future_extension.dart
import 'dart:async';

import '../../funcy.dart';

/// [Future]-[Option] monad transformer extension.
extension FutureOptionTransformerExtension<T> on Future<Option<T>> {
  /// Gets "inside" the future and performs [map] on inner [Option].
  Future<Option<R>> map<R>(R Function(T) f) => then((opt) => opt.map(f));

  /// Gets "inside" the future and performs [bind] on inner [Option].
  Future<Option<R>> bind<R>(FutureOr<Option<R>> Function(T) f) async {
    ArgumentError.checkNotNull(f, 'f');
    final maybeT = await this;
    if (maybeT is Some<T>) {
      return f(maybeT.value);
    } else {
      // ignore: prefer_const_constructors
      return None();
    }
  }
}

/// [Future]-[Either] monad transformer extension.
extension FutureEitherTransformerExtension<L, R> on Future<Either<L, R>> {
  /// Gets "inside" the future and performs [map] on inner [Either].
  Future<Either<L, R2>> map<R2>(R2 Function(R) f) =>
      then((either) => either.map(f));

  /// Gets "inside" the future and performs [bind] on inner [Either].
  Future<Either<L, R2>> bind<R2>(FutureOr<Either<L, R2>> Function(R) f) async {
    ArgumentError.checkNotNull(f, 'f');
    final lOrR = await this;
    if (lOrR is Right<L, R>) {
      return f(lOrR.right);
    } else {
      return Left(lOrR.leftOr(null));
    }
  }
}

/// [Future]-[Loadable] monad transformer extension.
extension FutureLoadableTransformerExtension<F, S> on Future<Loadable<F, S>> {
  /// Gets "inside" the future and performs [map] on inner [Loadable].
  Future<Loadable<F, R>> map<R>(R Function(S data) f) =>
      then((loadable) => loadable.map(f));

  /// Gets "inside" the future and performs [bind] on inner [Loadable].
  Future<Loadable<F, R>> bind<R>(
      FutureOr<Loadable<F, R>> Function(S data) f) async {
    ArgumentError.checkNotNull(f, 'f');
    final loadableS = await this;
    if (loadableS is Success<F, S>) {
      return f(loadableS.data);
    } else if (loadableS is Failed<F, S>) {
      return Failed(loadableS.failure);
    } else {
      return const Loading();
    }
  }
}


example/example.dart
// ignore_for_file: avoid_print
import 'package:funcy/funcy.dart';
import 'package:meta/meta.dart';

void main() {
  final employeeJson = retrieveEmployeeJson();
  final maybeEmployee =
      Employee.tryFromJson(employeeJson) | tryFallbackEmployee();
  maybeEmployee.option(
    () => print('Something really bad is happening '
        'even fallback employee does not work.'),
    (employee) => print('${employee.name} is ${employee.age}.'),
  );
}

Map<String, dynamic> retrieveEmployeeJson() => <String, dynamic>{
      'name': 'Bob',
      'position': 'boss',
      'age': 42,
    };

Option<Employee> tryFallbackEmployee() => Some(
      Employee(
        name: 'Fallback Employee',
        position: Position.subsubboss,
        age: -1,
      ),
    );

enum Position {
  boss,
  subboss,
  subsubboss,
}

Option<Position> tryParsePosition(String positionStr) {
  switch (positionStr) {
    case 'boss':
      return const Some(Position.boss);
    case 'subboss':
      return const Some(Position.subboss);
    case 'subsubboss':
      return const Some(Position.subsubboss);
    default:
      return const None();
  }
}

class Employee {
  Employee({
    @required this.name,
    @required this.position,
    @required this.age,
  });

  final String name;
  final Position position;
  final int age;

  static Option<Employee> tryFromJson(Map<String, dynamic> json) {
    return json.tryGet<String>('name').bind(
      (name) {
        return json.tryGet<String>('position').bind(tryParsePosition).bind(
          (position) {
            return json.tryGet<int>('age').guard(_ageTest).map(
              (age) {
                return Employee(
                  name: name,
                  position: position,
                  age: age,
                );
              },
            );
          },
        );
      },
    );
  }

  static bool _ageTest(int age) => age >= 18;
}
